### ISSUE TITLE
problem about inference

### ISSUE THREAD

**Issue by linjikun1** [CupCupRay/Bin2Summary#4]
Hi,

- How did you get raw_comment.ref? I downloaded the [complete dataset](https://drive.google.com/drive/folders/1ZgmuB5T2Ho5NRmJLozpUcmWvsEt3YxUT). But It seems that because the function name obtained by binaryninja from the raw_binaries is almost impossible to match related func in the raw_comment.ref file, or some reasons else, only 7 related `this_com` were found in the train set, 3 related `this_com` were found in the val set and 1 related `this_com` were found in the test set, according to [this function](https://github.com/CupCupRay/Bin2Summary/blob/a6afb97903c12d674bc3c6073ec36f0a51e9461d/src/palmtree_emb/gen_asm.py#L288). This results in very few subsequent datasets for training the Seq2Seq model.

  > Preparing the embedding data in ./data/train_
  Preparing the embedding data in ./data/val_
  Preparing the embedding data in ./data/test_
  And 7 <comment, snippet> pairs for training.
  And 3 <comment, snippet> pairs for validation.
  And 1 <comment, snippet> pairs for testing.

  And of those 7 data pairs for training, even duplication lead to only two data pairs in fact.

  ![Screenshot from 2024-11-09 16-24-27](https://github.com/user-attachments/assets/4f14e238-25b9-457d-b2ac-085192cfe27f)

  I'm not sure if it's normal. And I'm also not sure if it's a problem with `find_related_comment` function, or with the overly complex func_name generated by BinaryNinja, or with the raw_comment.ref dataset.

- When train and test a Seq2Seq model for comment generation, an error occured around Epoch 30/1000:

  > . Consider either turning off auto-sharding or switching the auto_shard_policy to DATA to shard this dataset. You can do this by creating a new tf.data.Options() object then setting options.experimental_distribute.auto_shard_policy = AutoShardPolicy.DATA before applying the options object to the dataset via dataset.with_options(options).
  1/1 [==============================] - 0s 80ms/step
  2024-11-09 03:33:20.279324: W tensorflow/core/grappler/optimizers/data/auto_shard.cc:776] AUTO sharding policy will apply DATA sharding policy as it failed to apply FILE sharding policy because of the following reason: Did not find a shardable source, walked to a node which is not a dataset: name: "FlatMapDataset/_9"

  And I'm not sure if this is due to a small dataset I mentioned above or not.

- My goal is to apply your method for binary function comment generation for other binary datasets. So I would like to ask whether or not you can share the pre-trained assembly language model and pre-trained Seq2Seq model so that I could do some inference directly?

Please help me.

**Comment by linjikun1** [comment]
I specifically print the `name` [here](https://github.com/CupCupRay/Bin2Summary/blob/a6afb97903c12d674bc3c6073ec36f0a51e9461d/src/palmtree_emb/gen_asm.py#L287), and the following snippets of results seem to indicate that the func_name which BinaryNinja gets from raw_binaries is readable.

> name = noise_ultralight
name = so_peercred
name = printer_start_job
name = printer_job_data
name = printer_finish_job
name = localproxy_fromfd_cmp
name = localproxy_fromfd_find
name = localproxy_tofd_cmp
name = localproxy_tofd_find
name = localproxy_errfd_cmp
name = localproxy_errfd_find
name = sk_localproxy_plug
name = localproxy_try_send

I guess maybe the raw_comment.ref file needs more attention. So could you tell me more about raw_comment.ref?

**I sincerely appreciate your guidance.**

**Comment by CupCupRay** [comment]
Hi, @ljk419511,

Sorry for the late reply. 

There is definitely something wrong if you only get so few samples in train/val/test datasets. In our evaluation, we have about 30K functions collected with good comments for reference (even in the coreutils, there are more than thousands of functions).

After seeing your question, I did a small test, which seems to work fine on my server. For example, in the file chmod.exe, I tried to get one sample, which is found in raw_comment.ref:

> Name: quotearg_buffer; Comment: Place into buffer BUFFER (of size BUFFERSIZE) a quoted version of argument ARG (of size ARGSIZE), using O to control quoting. If O is null, use the default. Terminate the output with a null character, and return the written size of the output, not counting the terminating null. If BUFFERSIZE is too small to store the output string, return the value that would have been returned had BUFFERSIZE been large enough. If ARGSIZE is SIZE_MAX, use the string length of the argument for ARGSIZE. 

I guess maybe there is something that does not match BinaryNinja and scripts. For example, some versions may collect the function with the name "fcn.quotearg_buffer", so you can feel free to edit my scrips in such scenarios.

Best regards,
Ray
